Split the entry-point-generation phase into two parts

smowton opened the pull request:
The new generate_support_functions phase creates __CPROVER_start and __CPROVER_initialize and runs before lazy loading; the final phase will run *after* lazy loading, and is therefore used to patch up changes that require functions' bodies to have been loaded already.

In subsequent PRs, lazy loading will take advantage of this to read __CPROVER_start and use it to guide its work; CI-lazy-loading will use it to replace its current hackish logic to guess how globals and parameters will be initialised.

smowton committed:
Split the entry-point-generation phase into two parts

The new generate_support_functions phase creates __CPROVER_start and
__CPROVER_initialize and runs before lazy loading; the final phase will run
*after* lazy loading, and is therefore used to patch up changes that require
functions' bodies to have been loaded already.

reuk commented on src/util/language.h:
```diff
@@ -45,6 +45,17 @@ class languaget:public messaget
+  /// complete. Functions introduced here are visible to lazy loading and
+  /// can influence its decisions (e.g. picking the types of input parameters
+  /// and globals), whereas anything introduced during `final` cannot.
+  virtual bool generate_support_functions(
```
Should probably be `private`.

reuk reviewed (changes requested):
Virtual methods used to customize behavior should generally be private.

thk123 commented on src/ansi-c/ansi_c_language.cpp:
```diff
@@ -117,6 +117,21 @@ bool ansi_c_languaget::parse(
   return result;
 }
 
+bool ansi_c_languaget::generate_support_functions(
```
Consider adding doxygen documentation for what supporting functions this generates (opaque functions, entry point)

smowton replied:
I have this in `languaget`, is that enough or does it have to be duplicated in derived classes?

thk123 commented on src/goto-programs/initialize_goto_model.cpp:
```diff
@@ -117,9 +118,11 @@ bool initialize_goto_model(
+        if(language_files.generate_support_functions(goto_model.symbol_table))
         {
-          msg.error() << "CONVERSION ERROR" << messaget::eom;
+          msg.error() << "ENTRY-POINT GENERATION ERROR" << messaget::eom;
```
Though most of `generate_support_functions` is related to the entry point, in might be worth calling this `SUPPORT FUNCTION ERROR`. Alternatively, could have the `generate_support_functions` throw structured exceptions for error states?

peterschrammel replied:
Structured exception will come, but let's do that separately.

thk123 reviewed (approved):
Seems reasonable - some suggestions if you agree with them. One thing to bare in mind is that the language appears to get thrown away and recreated (meaning the cmdline isn't parsed to the new instance). I'm going to create a PR that basically `assert`s that this is called _before_ calling any of the methods. You  might want to rebase this ontop of that and add the assert to `generate_support_functions` if it gets in first...

mgudemann commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -385,26 +407,9 @@ void java_bytecode_languaget::replace_string_methods(
-      max_nondet_array_length,
-      max_nondet_tree_depth,
-      get_pointer_type_selector());
+  return false;
```
Would you mind adding a comment to this?
Sometime later it might not be obvious why this cannot fail.

smowton commented:
@thk123 suggest checking https://github.com/diffblue/security-scanner/pull/192/files for the latest version of this (will update that one once this passes)--

Changes since this version:
* Removed `generate_entry_point`, replaced with a call to `generate_support_functions` (which must work correctly if run more than once; amended the Java version to match the Ansi-C version in this respect)
* Only use `rebuild_goto_start_function` when the binaries provided one and we're directed to overwrite it; otherwise run the `language_filest` version which searches for a frontend that can provide an entry-point.
* Move declaration of the `--function` option into `language.h`, which is I think its proper owner

thk123 commented:
@smowton Sorry your last message confused me - which PR should I be reviewing, this on diffblue/security-scanner#192?

peterschrammel commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -629,20 +629,6 @@ int cbmc_parse_optionst::get_goto_program(
     // are already compiled
       return 6;
 
-    if(cmdline.isset("function"))
```
Why is this removed? What replaces this functionality?

smowton replied:
This was a duplicate of functionality already in `initialize_goto_model` -- it was previously overwriting the start function twice.

smowton committed:
Add tests for mixed GOTO and C input

smowton commented:
@thk123 please review diffblue/security-scanner#192 then I once that's in I will clone it into this one for submission to develop

peterschrammel commented on src/util/language_file.cpp:
```diff
@@ -91,6 +91,22 @@ bool language_filest::parse()
+{
+  std::set<std::string> languages;
+
+  for(file_mapt::iterator it=file_map.begin();
```
ranged for?

smowton replied:
ok

thk123 commented:
I've reviewed 192 - once you've updated this PR to mirror could you ping me so I can create a TG bump since this will probably require some changes (I overhead you mention about the TG specific version of the language)

smowton committed:
Switch to C++11 for-loops in language_file.cpp

smowton commented:
@thk123 @peterschrammel this has been updated; Peter your suggested C++11 loops have been added as a separate commit.

thk123 commented:
I am creating a TG bump for this to verify it - but then can be merged

smowton commented:
@thk123 please press the button when you're happy

thk123 reviewed (changes requested):
For those out of ear shot of me and @smowton - there was a problem merging in to TG which @smowton is looking at at the moment.

smowton commented:
Updated; created companion test-gen PR https://github.com/diffblue/test-gen/pull/1085 to verify this doesn't break it

thk123 commented:
TG bump passes so have set approval.
